policy-controller/src/main.rs

```diff
@@ -113,75 +118,9 @@ async fn main() -> Result<()> {
-        let listen_addr = tcp.local_addr()?;
-        // Connection accept loop
-        let server = async move {
-            let http = hyper::server::conn::Http::new();
-            let handler = admission::Service { client };
-
-            loop {
-                let socket = match tcp.accept().await {
-                    Ok((socket, _)) => socket,
-                    Err(err) => {
-                        error!(%err, "Failed to accept connection");
-                        continue;
-                    }
-                };
-                let tls_cfg = {
-                    // Load public certificate.
-                    let certs = match load_certs("/var/run/linkerd/tls/tls.crt")
-                        .await
-                        .with_context(|| "failed to load certificate")
-                    {
-                        Ok(certs) => certs,
-                        Err(error) => {
-                            error!(%error);
-                            continue;
-                        }
-                    };
-                    // Load private key.
-                    let key = match load_private_key("/var/run/linkerd/tls/tls.key")
-                        .await
-                        .with_context(|| "failed to load private key")
-                    {
-                        Ok(key) => key,
-                        Err(error) => {
-                            error!(%error);
-                            continue;
-                        }
-                    };
-                    // Do not use client certificate authentication.
-                    let mut cfg = match rustls::ServerConfig::builder()
-                        .with_safe_defaults()
-                        .with_no_client_auth()
-                        .with_single_cert(certs, key)
-                    {
-                        Ok(cfg) => cfg,
-                        Err(error) => {
-                            error!(%error, "Failed to configure TLS");
-                            continue;
-                        }
-                    };
-                    // Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
-                    cfg.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
-                    sync::Arc::new(cfg)
-                };
-                let tls_acceptor = TlsAcceptor::from(tls_cfg);
-                let stream = match tls_acceptor.accept(socket).await {
-                    Ok(stream) => stream,
-                    Err(e) => {
-                        error!(%e, "TLS Error");
-                        continue;
-                    }
-                };
-                let conn = http.serve_connection(stream, handler.clone());
-                tokio::spawn(conn);
-            }
-        };
-
+        let (listen_addr, serve) = bind_admission_controller(bind_addr, client).await?;
+        tokio::spawn(serve.instrument(info_span!("admission")));
```

Comment:
better to keep all of these details out of the `main` so that both parts are easier to read in isolation